doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,355 @@
+Reconfiguration replicas will use the `reconfig` prefix; ex `reconfig-r3`,
+`reconfig-r4`, which will be removed once they are moved to active.
+
+
+### Implementation Details
+We want to accomplish the following
+ - interpret SQL to identify when to use graceful vs disruptive alters
+ - selectively choose which mechanism is used to alter the cluster
+ - provide a mechanism to cancel the operation
+
+
+#### Catalog Changes
+We should add a `pending: bool` field to the `ClusterReplica` memory and durable
+catalog structs. This will be set to true for reconfiguration replicas and will
+help either recover or cleanup on environmentd crashes. We should also propagate
+this to `mz_cluster_replicas`.
+
+#### Process
+Below broken this process broken down into four components
+1. Interperet the query to determine if we're doing graceful reconfiguration
+2. run `sequence_alter_cluster_managed_to_managed_graceful`
+3. wait and finalize
+4. return to the user
+
+*In v2, steps 3 and 4 may be swapped.
+
+__Interpreting the Query__
+This will use standard parsing/planning, the only notable
+callout here is that we'll want `mz_sql::plan::AlterClusterOptions`
+to contain an `alter_mechanaism` field which holds an optional
+`AlterClusterMechanism` which will look something like:
+
+```rust
+enum AlterClusterMechanism {
+  Duration{
+    timeout: Duration
+    ...
+  },
+  UntilCaughtUp{
+    timeout: Duration,
+    continue_on_timeout: bool
+    ...
+  },
+}
+```
+
+__Cluster Sequencer__
+
+The following roughly outlines the new logic for the cluster sequencer when
+performing an alter cluster on a managed cluster.
+
+We'll need to look at the alter_mechanism to determine if
+we need to run additional validations and whether to use the
+`sequence_alter_cluster_managed_to_managed_disruptive` or
+`sequence_alter_cluster_managed_to_managed_graceful` function.
+
+
+The following are validations to run if the `AlterClusterPlan` contains
+an alter_mechanism.
+1. validate the cluster is managed
+2. validate that the cluster is not undergoing a reconfiguration
+3. validate the cluster has no sources and now sinks
+4. the cluster cannot have a refresh schedule (v1)
+
+
+`sequence_alter_cluster_managed_to_managed_disruptive` will  perform the current
+delete-before-create mechanism.
+
+
+`sequence_alter_cluster_managed_to_managed_graceful`  will still apply
+all limits and validity checks to the `new configuration`, but will not
+perform deletes of the active replicas, nor commit the new config to the
+catalog. Instead, it will ask the compute controller to create new replicas,
+add a new entry in  `pending_cluster_reconfigurations`,  then kick off a
+`ClusterReconfigureCheckReady` message.
+
+The state for the reconfiguration will largely be stored in a `PendingClusterReconfiguration`
+struct.
+
+
+```rust
+struct PendingClusterReconfiguration {
+ ctx: ExecuteContext,
+ // used to determine if the message_handler
+ // should wait, or finalize
+ is_ready: bool
```

Comment:
Possibly this is not needed. It might be able to be an impl method on the struct.

Reply:
What do you think about having the `handle_cluster_reconfiguration` message handler  always validate the check and re-sen the message if it's false. We wouldn't need to keep track of is ready at all.
Ex:
```
if the start_time.elapsed > wait_time {
   //finalize
 } else {
         mz_ore::task::spawn(|| "", async move {
            // some wait condition
            tokio::time::sleep(wait_time - start_time.elapsed).wait;
            let _ = internal_cmd_tx.send(Message::ClusterReconfigurationReady(connid));
        });
 } 
 ```
 
 This makes more sense in the `WAIT UNTIL CAUGHT UP` case, where, if not caught up, it would just send the message in 3 seconds.

Reply:
Yes, this is good.

Reply:
coool updated!